### PR TITLE
OIDC login workflow

### PR DESCRIPTION
**Pull request by Yibaebi**

### CODE REVIEW

File: app/services/oidc.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    }
+  });
+
+  async validateOidcTokenOrRedirect(token: string) {
+    if (this.session.isAuthenticated) {
+      await this.validateOidcToken.perform(token);
+    } else {
+      this.router.transitionTo('login');
+
+      this.window.sessionStorage.setItem('oidc_token', token);
```
Comment (cosmosgenius) [inline app/services/oidc.ts:93]: Shouldn't this be first logically before initiating login redirect?

**Review by cosmosgenius — CHANGES_REQUESTED** [review]

File: app/services/oidc.ts (hunk 1 above)
Comment (future-pirate-king) [reply app/services/oidc.ts:93]: it doesn't matter `oidc_token` is required after login

File: app/services/oidc.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+  redirect_url: null | string;
+  error: null | {
+    code: string;
+    description: string;
+  };
+}
+
+type ValidateOidcTokenResponse = OidcResponse;
+
+type OidcAuthorizeResult = OidcResponse;
```
Comment (Yibaebi) [inline app/services/oidc.ts:19]: Are the `ValidateOidcTokenResponse`, `OidcValidationResult`, and `OidcAuthorizeResult` types required?

File: app/authenticators/irene.js [hunk 3]
```diff
@@ -18,55 +18,63 @@ const processData = (data) => {
...
-    return ajax.post(url, { data })
-      .then(data => {
-        data = processData(data);
-        this.resumeTransistion();
-        return data;
-      });
+
+    return ajax.post(url, { data }).then((data) => {
+      data = processData(data);
+
```
[the remark this answers is not in the document]
Comment (future-pirate-king) [reply app/authenticators/irene.js:51]: currently this is not having any effect on login flow

File: app/authenticators/saml2.js [hunk 4]
```diff
@@ -1,39 +1,55 @@
...
-        return reject(msg);
-      });
+
+      this.get('ajax')
+        .post(url, { data: { token: ssotoken } })
+        .then(
+          (data) => {
+            data = processData(data);
+
+            resolve(data);
```
[the remark this answers is not in the document]
Comment (future-pirate-king) [reply app/authenticators/saml2.js:33]: currently this is not having any effect on login flow

File: app/services/oidc.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+          },
+        });
+
+        this.window.sessionStorage.removeItem('oidc_token');
+      }
+    }
+  }
+
+  authorizeOidcAppPermissions = task(async (token: string, allow?: boolean) => {
+    const res = await this.network.post(this.oidcAuthorizeEndpoint, {
```
[the remark this answers is not in the document]
Comment (future-pirate-king) [reply app/services/oidc.ts:63]: it has been handled below

File: app/services/oidc.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    if (this.session.isAuthenticated) {
+      await this.validateOidcToken.perform(token);
+    } else {
+      this.router.transitionTo('login');
+
+      this.window.sessionStorage.setItem('oidc_token', token);
+    }
+  }
+
+  validateOidcToken = task(async (token: string) => {
```
[the remark this answers is not in the document]
Comment (future-pirate-king) [reply app/services/oidc.ts:97]: it has been handled below

**Pull request merged**